src/hotspot/share/gc/parallel/psOldGen.cpp

```diff
@@ -395,7 +395,7 @@ class VerifyObjectStartArrayClosure : public ObjectClosure {
     _start_array(start_array) { }
 
   virtual void do_object(oop obj) {
-    HeapWord* test_addr = cast_from_oop<HeapWord*>(obj) + 1;
+    HeapWord* test_addr = cast_from_oop<HeapWord*>(obj);
```

Comment:
I thought this `+1` is specifically to test that `object_start` is able to find the object header when given the interior pointer. See the `guarantee`-s in the next lines.

Reply:
Yes, but with compact headers, we could now have 1-word-sized objects, in which case this would fail. I am not sure how to deal with that, TBH. Maybe do the whole test only when !UseCompactObjectHeaders or when object-size is > 1?

Reply:
I think we should keep the original code semantics intact, and `guarantees` might need to check stuff even with zero offset.

```
// With compact headers, the objects can be one-word sized.
size_t int_off = UseCompactObjectHeaders ? MIN2(1, obj->size() - 1) : 1;
HeapWord* test_addr = cast_from_oop<HeapWord*>(obj) + int_off;
```